In [1]:
pip install pdfplumber

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 kB 1.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.5/48.5 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.0/60.0 kB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 53.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.8/2.8 MB 75.5 MB/s eta 0:00:00


In [3]:
import pdfplumber
import pandas as pd

In [4]:
'''
df_pdf = pp.open("/content/drive/MyDrive/mapinskhu/[25-2]Data/수강신청안내책자-20250821-crop.pdf")

# pdf의 필요한 테이블을 데이터프레임으로 저장

In [17]:
turn = [
    [1, 2], [1, 2], [1], [1], [1, 2], [1, 2], [1, 2], [1, 2], [1, 2, 3], [1, 2, 3],
    [1, 2], [1], [1, 2], [1, 2, 3], [1, 2], [1], [1], [1], [1], [1], [1], [1], [1],
    [1], [1], [1], [1], [1], [1], [2], [2], [1], [1], [1], [1], [2], [1]
]
tables = []

with pdfplumber.open("/content/drive/MyDrive/mapinskhu/[25-2]Data/수강신청안내책자-20250821-crop.pdf") as pdf:
    # PDF 페이지 수가 turn 리스트 길이보다 적은 경우를 대비하여
    # 더 작은 값으로 루프 횟수를 제한합니다.
    num_pages_to_process = min(len(turn), len(pdf.pages))

    for i in range(num_pages_to_process):
        # 해당 페이지에 존재하는 모든 테이블 추출
        page_tables = pdf.pages[i].extract_tables()

        # turn에 명시된 모든 테이블 번호를 순회합니다.
        for t in turn[i]:
            # 추출하려는 테이블 번호(t-1)가 실제 존재하는지 확인합니다.
            if t > 0 and t <= len(page_tables):
                tables.append(pd.DataFrame(page_tables[t - 1]))
            else:
                print(f"경고: {i+1} 페이지에 {t}번째 테이블이 존재하지 않습니다.")

# 추출된 모든 테이블을 하나의 파일로 저장하거나
# tables 리스트를 활용하여 원하는 작업을 수행할 수 있습니다.
# 예: tables[0].to_excel("first_table.xlsx")

In [5]:
# 페이지 마다 추출할 테이블 번호 저장
'''
turn = [
    [1,2],[1,2],[1],[1],[1,2],[1,2],[1,2],[1,2],[1,2,3],[1,2,3],
    [1,2],[1],[1,2],[1,2,3],[1,2],[1],[1],[1],[1],[1],[1],[1],[1],
    [1],[1],[1],[1],[1],[1],[2],[2],[1],[1],[1],[1],[2],[1]
]
print(len(turn))

37


In [6]:
# 총 테이블 개수 확인
'''
cnt = 0
for i in range(len(turn)):
    for j in range(len(turn[i])):
        cnt+=1

print(cnt, turn)

52 [[1, 2], [1, 2], [1], [1], [1, 2], [1, 2], [1, 2], [1, 2], [1, 2, 3], [1, 2, 3], [1, 2], [1], [1, 2], [1, 2, 3], [1, 2], [1], [1], [1], [1], [1], [1], [1], [1], [1], [1], [1], [1], [1], [1], [2], [2], [1], [1], [1], [1], [2], [1]]


In [7]:
# 테이블을 데이터프레임으로 저장
'''
tables= []
for i in range(len(turn)):
    # 해당 페이지에 존재하는 모든 테이블 추출
    page_tables = df_pdf.pages[i].extract_tables()
    #print(type(page_tables)) -> list
    for t in turn[i]:
        tables.append(pd.DataFrame(page_tables[t-1]))

In [18]:
len(tables)

52

# 데이터프레임 정리
##필요한 컬럼만 남기기

In [19]:
print(tables[18])

   0       1       2        3   4             5   6    7       8       9   \
0  순번  이수\n구분  개설\n학년     과목코드  분반          교과목명  학점  교수명  학년\n제한  전공\n제한   
1   1      교선      전체  AH00022  01    진로탐색과 자기계발   2  한상진       N       N   
2   2      교선      전체  AH00056  01      창업과 지식재산   2  류선미       N       N   
3   3      교선      전체  AH00057  01  스타트업비즈니스전략수립   3  최우진       N       N   
4   4      교선      전체  AK00153  01  직업으로서의시민사회활동   3  박경주       N       N   

                   10    11  
0              강의실/시간  확인사항  
1  7208/화/15:00~16:50        
2  1503/화/10:00~11:50        
3  7305/월/12:00~14:50        
4  3501/화/12:00~14:50        


In [20]:
# 컬럼명 변경
for i, table in enumerate(tables):
    tables[i].columns = table.iloc[0,:]
    tables[i] = table.iloc[1:, :]

In [21]:
print(tables[18])

0 순번 이수\n구분 개설\n학년     과목코드  분반          교과목명 학점  교수명 학년\n제한 전공\n제한  \
1  1     교선     전체  AH00022  01    진로탐색과 자기계발  2  한상진      N      N   
2  2     교선     전체  AH00056  01      창업과 지식재산  2  류선미      N      N   
3  3     교선     전체  AH00057  01  스타트업비즈니스전략수립  3  최우진      N      N   
4  4     교선     전체  AK00153  01  직업으로서의시민사회활동  3  박경주      N      N   

0              강의실/시간 확인사항  
1  7208/화/15:00~16:50       
2  1503/화/10:00~11:50       
3  7305/월/12:00~14:50       
4  3501/화/12:00~14:50       


In [22]:
print(tables[15])

0 순번 이수\n구분 개설\n학년     과목코드  분반                   교과목명 학점  교수명 학년\n제한 전공\n제한  \
1  1     교선     전체  AD00033  01                 대학합창II  3  차피득      N      N   
2  2     교선     전체  AD00193  01  실용음악의 이해와 감상Ⅱ\n(대중음악)  2  하지원      N      N   

0              강의실/시간 확인사항  
1  9301/수/15:00~17:50       
2  3402/화/15:00~16:50       


In [7]:
# 각 테이블 컬럼 확인하기
# 공통 :
# ['순번', '이수\n구분', '개설\n학년', '과목코드', '분반', '교과목명', '학점', '교수명', '강의실/시간', '확인사항']
'''
for i in tables:
  print(i.columns)

Index(['순번', '이수\n구분', '개설\n학년', '과목코드', '분반', '교과목명', '학점', '교수명', '학년\n제한',
       '전공\n제한', '강의실/시간', '확인사항'],
      dtype='object', name=0)
Index(['순번', '이수\n구분', '개설\n학년', '과목코드', '분반', '교과목명', '학점', '교수명', '학년\n제한',
       '전공\n제한', '강의실/시간', '확인사항'],
      dtype='object', name=0)
Index(['순번', '이수\n구분', '개설\n학년', '과목코드', '분반', '교과목명', '학점', '교수명', '학년\n제한',
       '전공\n제한', '강의실/시간', '확인사항'],
      dtype='object', name=0)
Index(['순번', '이수\n구분', '개설\n학년', '과목코드', '분반', '교과목명', '학점', '교수명', '학년\n제한',
       '전공\n제한', '강의실/시간', '확인사항'],
      dtype='object', name=0)
Index(['순번', '이수\n구분', '개설\n학년', '과목코드', '분반', '교과목명', '학점', '교수명', '학년\n제한',
       '전공\n제한', '강의실/시간', '확인사항'],
      dtype='object', name=0)
Index(['순번', '이수\n구분', '개설\n학년', '과목코드', '분반', '교과목명', '학점', '교수명', '학년\n제한',
       '전공\n제한', '강의실/시간', '확인사항'],
      dtype='object', name=0)
Index(['순번', '이수\n구분', '개설\n학년', '과목코드', '분반', '교과목명', '학점', '교수명', '학년\n제한',
       '전공\n제한', '강의실/시간', '확인사항'],
      dtype='object', 

In [23]:
# 컬럼 존재 확인 후 필요한 컬럼만 남기기

# 원하는 컬럼명의 리스트
desired_columns = ['이수\n구분', '과목코드', '교과목명', '교수명', '강의실/시간', '확인사항']

# 각 테이블을 순회하며 컬럼명을 확인
for i, table_df in enumerate(tables):
    # 데이터프레임의 컬럼명을 리스트로 가져옵니다.
    current_columns = list(table_df.columns)

    # desired_columns의 모든 요소가 current_columns에 포함되어 있는지 확인
    # all() 함수와 set()을 사용하여 효율적으로 검사합니다.
    # set()의 in 연산은 리스트보다 빠르므로 대량의 데이터 처리 시 유리합니다.
    if set(desired_columns).issubset(set(current_columns)):
        print(f"{i+1}번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅")
        tables[i] = table_df[desired_columns]
    else:
        # 어떤 컬럼이 누락되었는지 확인
        missing_columns = list(set(desired_columns) - set(current_columns))
        print(f"{i+1}번째 테이블: 원하는 컬럼 중 일부가 누락되었습니다. ❌ -> 누락된 컬럼: {missing_columns}")

1번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
2번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
3번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
4번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
5번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
6번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
7번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
8번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
9번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
10번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
11번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
12번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
13번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
14번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
15번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
16번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
17번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
18번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
19번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
20번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
21번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
22번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
23번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
24번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
25번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
26번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
27번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
28번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
29번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
30번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
31번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
32번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
33번째 테이블: 모든 원하는 컬럼이 존재합니다. ✅
34번째 테이블: 모든 원하는 컬럼

In [13]:
# 필요한 컬럼만 남기기
'''
for i, table in enumerate(tables):
    # tables[i] = table.loc[:, [1,3,5,7,10,11]]
    tables[i] = table[['이수\n구분', '과목코드', '교과목명', '교수명', '강의실/시간', '확인사항']]
    # tables[i] = table.loc[:, ['이수\n구분','과목코드','교과목명','교수명','강의실/시간','확인사항']]

In [28]:
print(tables[17])

0 이수\n구분     과목코드                 교과목명   교수명              강의실/시간  \
1     교선  AE00029               사회봉사II   김연아                       
2     교선  AK00155  비영리와사회혁신:\n필란트로피리더십   이선희  3402/목/14:00~16:50   
3     교선  AK00080         대학생사회공헌프로젝트2   김연아  7202/금/15:00~17:50   
4   None     None                 None  None                None   
5   None     None                 None  None                None   
6   None     None                 None  None                None   
7   None     None                 None  None                None   

0                              확인사항  
1  출석수업 없음-\n사회봉사 필수 교과목 이수자만\n수강가능  
2    필란트로피의 이해와 실천[AE00092]\n과목명 변경  
3         1) 사회봉사 필수 교과목 이수자만\n수강가능  
4                                    
5                                    
6                                    
7                                    


In [24]:
print(tables[18])

0 이수\n구분     과목코드          교과목명  교수명              강의실/시간 확인사항
1     교선  AH00022    진로탐색과 자기계발  한상진  7208/화/15:00~16:50     
2     교선  AH00056      창업과 지식재산  류선미  1503/화/10:00~11:50     
3     교선  AH00057  스타트업비즈니스전략수립  최우진  7305/월/12:00~14:50     
4     교선  AK00153  직업으로서의시민사회활동  박경주  3501/화/12:00~14:50     


# 테이블 합치기

In [12]:
# 컬럼 헤더 지정
'''
header = ['kind','code','classname','prof','roomntime','check']

In [17]:
# 테이블 합치기
'''
tablecopy = tables.copy()
df = tablecopy[0]
for t in tablecopy[1:]:
    df = pd.concat([df, t], axis=0)

In [25]:


# 1. 10개 컬럼으로 구성된 테이블 리스트 (tables)가 있다고 가정
#    tables = [pd.DataFrame(...), pd.DataFrame(...), ...]

# 2. 필요한 6개의 컬럼과 새로운 헤더를 정의합니다.
#    (이전 컬럼명과 새 컬럼명을 순서대로 대응시켰습니다.)
selected_columns = ['이수\n구분', '과목코드', '교과목명', '교수명', '강의실/시간', '확인사항']
new_headers = ['kind', 'code', 'classname', 'prof', 'roomntime', 'check']

# 3. 모든 테이블을 순회하며 컬럼을 정리합니다.
processed_tables = []
for i, table_df in enumerate(tables):
    try:
        # 10개 컬럼 중 원하는 6개 컬럼만 선택
        filtered_df = table_df[selected_columns]

        # 선택된 컬럼들의 이름을 새로운 헤더로 변경
        filtered_df.columns = new_headers

        # 정리된 테이블을 새로운 리스트에 추가
        processed_tables.append(filtered_df)
        print(f"{i+1}번째 테이블: 컬럼 정리가 완료되었습니다. ✅")
    except KeyError as e:
        # 컬럼을 찾을 수 없으면 오류 메시지를 출력하고 건너뜁니다.
        print(f"오류: {i+1}번째 테이블에서 필요한 컬럼을 찾을 수 없습니다. -> {e}")

# 4. 컬럼명이 통일된 테이블들을 합칩니다.
df = pd.concat(processed_tables, ignore_index=True)


1번째 테이블: 컬럼 정리가 완료되었습니다. ✅
2번째 테이블: 컬럼 정리가 완료되었습니다. ✅
3번째 테이블: 컬럼 정리가 완료되었습니다. ✅
4번째 테이블: 컬럼 정리가 완료되었습니다. ✅
5번째 테이블: 컬럼 정리가 완료되었습니다. ✅
6번째 테이블: 컬럼 정리가 완료되었습니다. ✅
7번째 테이블: 컬럼 정리가 완료되었습니다. ✅
8번째 테이블: 컬럼 정리가 완료되었습니다. ✅
9번째 테이블: 컬럼 정리가 완료되었습니다. ✅
10번째 테이블: 컬럼 정리가 완료되었습니다. ✅
11번째 테이블: 컬럼 정리가 완료되었습니다. ✅
12번째 테이블: 컬럼 정리가 완료되었습니다. ✅
13번째 테이블: 컬럼 정리가 완료되었습니다. ✅
14번째 테이블: 컬럼 정리가 완료되었습니다. ✅
15번째 테이블: 컬럼 정리가 완료되었습니다. ✅
16번째 테이블: 컬럼 정리가 완료되었습니다. ✅
17번째 테이블: 컬럼 정리가 완료되었습니다. ✅
18번째 테이블: 컬럼 정리가 완료되었습니다. ✅
19번째 테이블: 컬럼 정리가 완료되었습니다. ✅
20번째 테이블: 컬럼 정리가 완료되었습니다. ✅
21번째 테이블: 컬럼 정리가 완료되었습니다. ✅
22번째 테이블: 컬럼 정리가 완료되었습니다. ✅
23번째 테이블: 컬럼 정리가 완료되었습니다. ✅
24번째 테이블: 컬럼 정리가 완료되었습니다. ✅
25번째 테이블: 컬럼 정리가 완료되었습니다. ✅
26번째 테이블: 컬럼 정리가 완료되었습니다. ✅
27번째 테이블: 컬럼 정리가 완료되었습니다. ✅
28번째 테이블: 컬럼 정리가 완료되었습니다. ✅
29번째 테이블: 컬럼 정리가 완료되었습니다. ✅
30번째 테이블: 컬럼 정리가 완료되었습니다. ✅
31번째 테이블: 컬럼 정리가 완료되었습니다. ✅
32번째 테이블: 컬럼 정리가 완료되었습니다. ✅
33번째 테이블: 컬럼 정리가 완료되었습니다. ✅
34번째 테이블: 컬럼 정리가 완료되었습니다. ✅
35번째 테이블: 컬럼 정리가 완료되었습니다. ✅
36번째 테이블: 컬럼 정리가 완료되었습니다. ✅
3

In [27]:
df[110:120]
# 113~116 에 pdfplumber의 오류로 인해 None으로 추출되고 있음.
# 해결방법으로, dropna로  none을 삭제하면 됨.

kind     code            classname  prof           roomntime  \
110    교선  AE00029               사회봉사II   김연아                       
111    교선  AK00155  비영리와사회혁신:\n필란트로피리더십   이선희  3402/목/14:00~16:50   
112    교선  AK00080         대학생사회공헌프로젝트2   김연아  7202/금/15:00~17:50   
113  None     None                 None  None                None   
114  None     None                 None  None                None   
115  None     None                 None  None                None   
116  None     None                 None  None                None   
117    교선  AH00022           진로탐색과 자기계발   한상진  7208/화/15:00~16:50   
118    교선  AH00056             창업과 지식재산   류선미  1503/화/10:00~11:50   
119    교선  AH00057         스타트업비즈니스전략수립   최우진  7305/월/12:00~14:50   

                                check  
110  출석수업 없음-\n사회봉사 필수 교과목 이수자만\n수강가능  
111    필란트로피의 이해와 실천[AE00092]\n과목명 변경  
112         1) 사회봉사 필수 교과목 이수자만\n수강가능  
113                                    
114                                    
115                                    
116                                    
117                                    
118                                    
119

In [33]:

# 전체 컬럼 개수
total_columns = len(df.columns)

# 유효한 값이 2개 이하인 행을 제거 (즉, 3개 이상이 None인 행 제거)
cleaned_df = df.dropna(thresh=total_columns - 2)

# 2. 제거된 행으로 인해 불연속적으로 된 인덱스를 0부터 다시 시작하도록 재설정합니다.
#    'drop=True' 옵션은 기존 인덱스를 제거하고 새로운 인덱스를 만듭니다.
cleaned_df = cleaned_df.reset_index(drop=True)

In [34]:
cleaned_df[110:120]

kind     code            classname prof             roomntime  \
110   교선  AE00029               사회봉사II  김연아                         
111   교선  AK00155  비영리와사회혁신:\n필란트로피리더십  이선희    3402/목/14:00~16:50   
112   교선  AK00080         대학생사회공헌프로젝트2  김연아    7202/금/15:00~17:50   
113   교선  AH00022           진로탐색과 자기계발  한상진    7208/화/15:00~16:50   
114   교선  AH00056             창업과 지식재산  류선미    1503/화/10:00~11:50   
115   교선  AH00057         스타트업비즈니스전략수립  최우진    7305/월/12:00~14:50   
116   교선  AK00153         직업으로서의시민사회활동  박경주    3501/화/12:00~14:50   
117   교필  AK00113     비아메디아채플:그리스도교와세계  이종민  3502/월요일/12:00~12:50   
118   교필  AK00113     비아메디아채플:그리스도교와세계  이종민  3502/월요일/13:00~13:50   
119   교필  AK00113     비아메디아채플:그리스도교와세계  이종민  3502/월요일/14:00~14:50   

                                check  
110  출석수업 없음-\n사회봉사 필수 교과목 이수자만\n수강가능  
111    필란트로피의 이해와 실천[AE00092]\n과목명 변경  
112         1) 사회봉사 필수 교과목 이수자만\n수강가능  
113                                    
114                                    
115                                    
116                                    
117                                    
118                                    
119

In [17]:
# 헤더 설정
'''
df.columns = header

In [18]:
# 인덱스 초기화
'''
df.reset_index(inplace=True)

In [36]:
cleaned_df.head()

kind     code classname prof             roomntime              check
0   교필  AC00007     인권과평화  조경희  7207/화요일/10:00~11:50    주제 : 식민주의와 인종주의
1   교필  AC00007     인권과평화  강성현  7308/화요일/10:00~11:50        주제 : 전쟁과 평화
2   교필  AC00007     인권과평화  이상윤  7208/화요일/10:00~11:50  주제 : 세계시민의 비전과 도전
3   교필  AC00007     인권과평화  오영숙  3502/화요일/10:00~11:50            주제 : 미정
4   교필  AC00013       말과글  오현화  1503/수요일/09:00~10:50

# 이상한 행 추출

In [37]:
len(cleaned_df)

414

In [38]:
import numpy as np
cleaned_df[cleaned_df.classname.isna() | \
           cleaned_df.code.isna() | \
           cleaned_df.roomntime.isna()]

Empty DataFrame
Columns: [kind, code, classname, prof, roomntime, check]
Index: []

# 확인사항 컬럼 반영하기

In [40]:
record0 = cleaned_df.copy()
subjects = ['인권과평화', '대학생활세미나 II']

for index, row in record0.iterrows():
    if (row.classname in subjects):
        record0.loc[index, 'classname'] = row.classname + '(' + row.check + ')'

In [41]:
#check 열 삭제
record0 = record0.drop('check', axis=1)

# roomntime 컬럼 분리
### 1. 주2회 수업 분리 -> 주1회 수업은 rnt2에 NaN
### 2. 강의실/요일/시간 분리 -> room1, room2, date1, date2, start, end

In [42]:
record1 = record0.copy()

In [43]:
# 값에서 공백제거
record1["classname"] = record1["classname"].str.replace(" ", "")
record1["classname"] = record1["classname"].str.replace("\n", "")
record1["roomntime"] = record1["roomntime"].str.replace(" ", "")

### 1번 내용 분리

In [44]:
record2 = record1.copy()

In [45]:
record2['roomntime2'] = record2.roomntime.str.split('\n').str[1]
record2['roomntime'] = record2.roomntime.str.split('\n').str[0]

In [46]:
record2.tail(10)

kind     code  classname prof           roomntime          roomntime2
404   전선  IS00013      컴퓨터구조  정인철  3301/월/13:30~14:45  3301/수/13:30~14:45
405   전선  IS00014   JSP프로그래밍  박정식  6404/화/09:00~11:50                 NaN
406   전선  IS00015   백엔드프로그래밍  김선형  6202/화/15:00~16:15  6202/목/15:00~16:15
407   전선  IS00015   백엔드프로그래밍  김선형  6202/화/16:30~17:45  6202/목/16:30~17:45
408   전선  IS00016    컴퓨터네트워크  박정식  6406/목/09:00~11:50                 NaN
409   전선  IS00017    디지털회로실험  정인철  6406/월/09:00~11:50                 NaN
410   전선  IS00018  서버구축및형상관리  이승진  6202/월/15:00~16:15  6202/수/15:00~16:15
411   전선  IS00018  서버구축및형상관리  이승진  6202/월/16:30~17:45  6202/수/16:30~17:45
412   전선  IS00019     신호및시스템  김학수  6201/월/13:30~14:45  6201/수/13:30~14:45
413   전선  IS00020       웹디자인  이해신  3302/금/09:00~11:50                 NaN

In [47]:
# 주2회 수업중 각 수업의 시간과 강의실이 같은지 확인
# rnt2가 nan인 데이터 제외 추출
deleteNan = record2.dropna(axis=0, how='any', subset=['roomntime2'], inplace=False).copy()

In [48]:
deleteNan.head()

kind     code  classname       prof           roomntime  \
93    교선  AD00003      기초일본어  요코야마\n나호코  1505/월/13:30~14:45   
94    교선  AD00004      실용일본어  요코야마\n나호코  7208/월/12:00~13:15   
101   교선  AE00012   아시아사회와문화        박은홍  7206/화/16:30~17:45   
137   전탐  BH00005  네이티브일본어회화    요코야마나호코  7308/화/13:30~14:45   
138   전탐  BH00006   쉽게배우는중국어         장엽  7305/화/15:00~16:15   

             roomntime2  
93   1505/수/13:30~14:45  
94   7208/수/12:00~13:15  
101  7206/목/16:30~17:45  
137  7308/목/13:30~14:45  
138  7305/목/15:00~16:15

In [49]:
# room 또는 time이 일치하면 false 저장
deleteNan['testRoom'] = \
(deleteNan.roomntime.str.split('/').str[0]) \
!= (deleteNan.roomntime2.str.split('/').str[0])

deleteNan['testTime'] = \
(deleteNan.roomntime.str.split('/').str[2]) \
!= (deleteNan.roomntime2.str.split('/').str[2])

In [50]:
deleteNan.head()

kind     code  classname       prof           roomntime  \
93    교선  AD00003      기초일본어  요코야마\n나호코  1505/월/13:30~14:45   
94    교선  AD00004      실용일본어  요코야마\n나호코  7208/월/12:00~13:15   
101   교선  AE00012   아시아사회와문화        박은홍  7206/화/16:30~17:45   
137   전탐  BH00005  네이티브일본어회화    요코야마나호코  7308/화/13:30~14:45   
138   전탐  BH00006   쉽게배우는중국어         장엽  7305/화/15:00~16:15   

             roomntime2  testRoom  testTime  
93   1505/수/13:30~14:45     False     False  
94   7208/수/12:00~13:15     False     False  
101  7206/목/16:30~17:45     False     False  
137  7308/목/13:30~14:45     False     False  
138  7305/목/15:00~16:15     False     False

In [51]:
# true인 행 추출(==room또는 time이 다른것 추출)
deleteNan[deleteNan['testRoom']==True]

Empty DataFrame
Columns: [kind, code, classname, prof, roomntime, roomntime2, testRoom, testTime]
Index: []

### 2번 내용 분리

In [52]:
record2['room1'] = record2.roomntime.str.split('/').str[0]
record2['room2'] = record2.roomntime2.str.split('/').str[0]
record2['date1'] = record2.roomntime.str.split('/').str[1]
record2['date2'] = record2.roomntime2.str.split('/').str[1]
record2['start'] = record2.roomntime.str.split('/').str[2].str.split('~').str[0]
record2['end'] = record2.roomntime.str.split('/').str[2].str.split('~').str[1]

In [53]:
record2.drop(['roomntime', 'roomntime2'], axis=1, inplace=True) #필요없는 열 삭제

In [54]:
record2.head()

kind     code             classname prof room1 room2 date1 date2  start  \
0   교필  AC00007   인권과평화(주제:식민주의와인종주의)  조경희  7207   NaN   화요일   NaN  10:00   
1   교필  AC00007       인권과평화(주제:전쟁과평화)  강성현  7308   NaN   화요일   NaN  10:00   
2   교필  AC00007  인권과평화(주제:세계시민의비전과도전)  이상윤  7208   NaN   화요일   NaN  10:00   
3   교필  AC00007          인권과평화(주제:미정)  오영숙  3502   NaN   화요일   NaN  10:00   
4   교필  AC00013                   말과글  오현화  1503   NaN   수요일   NaN  09:00   

     end  
0  11:50  
1  11:50  
2  11:50  
3  11:50  
4  10:50

# 예외처리
## room1, start, end가 미지정인 경우 처리

In [55]:
# start end 없는 강의 삭제
import numpy as np
record2[record2.start.isna() | record2.end.isna()]
# df1[df1['start'] == np.nan | df1['end'] == np.nan].any()

kind     code          classname prof room1 room2         date1 date2  \
88    교선  AD00172  더불어숲인문학(삶을위한인문공부)  고병헌  7304   NaN  월09:00~11:50   NaN   
92    교선  AE00001            영상시대의이해  권승태  6110   NaN  월15:00~17:50   NaN   
110   교선  AE00029             사회봉사II  김연아         NaN           NaN   NaN   
288   전선  DJ00060                광고론  김인혜  7309   NaN             목   NaN   
290   전선  DJ00062              매스컴이론  이준형  7308   NaN             목   NaN   
323   전선  DP00008             웹드라마비평   강인  3501   NaN             목   NaN   
342   전선  XE00022               인턴십1  이상윤         NaN           NaN   NaN   
343   전선  XE00023               인턴십2  이상윤         NaN           NaN   NaN   
344   전선  XE00024               인턴십3  이상윤         NaN           NaN   NaN   
345   전선  XE00025               인턴십4  이상윤         NaN           NaN   NaN   
361   전선  XE00022               인턴십1  이상윤         NaN           NaN   NaN   
362   전선  XE00023               인턴십2  이상윤         NaN           NaN   NaN   
363   전선  XE00024               인턴십3  이상윤         NaN           NaN   NaN   
364   전선  XE00025               인턴십4  이상윤         NaN           NaN   NaN   
376   전선  XE00022               인턴십1  이상윤         NaN           NaN   NaN   
377   전선  XE00023               인턴십2  이상윤         NaN           NaN   NaN   
378   전선  XE00024               인턴십3  이상윤         NaN           NaN   NaN   
379   전선  XE00025               인턴십4  이상윤         NaN           NaN   NaN   
388   전선  XE00022               인턴십1  이상윤         NaN           NaN   NaN   
389   전선  XE00023               인턴십2  이상윤         NaN           NaN   NaN   
390   전선  XE00024               인턴십3  이상윤         NaN           NaN   NaN   
391   전선  XE00025               인턴십4  이상윤         NaN           NaN   NaN   

           start  end  
88           NaN  NaN  
92           NaN  NaN  
110          NaN  NaN  
288  12:00-14:50  NaN  
290  09:00-11:50  NaN  
323  09:00-11:50  NaN  
342          NaN  NaN  
343          NaN  NaN  
344          NaN  NaN  
345          NaN  NaN  
361          NaN  NaN  
362          NaN  NaN  
363          NaN  NaN  
364          NaN  NaN  
376          NaN  NaN  
377          NaN  NaN  
378          NaN  NaN  
379          NaN  NaN  
388          NaN  NaN  
389          NaN  NaN  
390          NaN  NaN  
391          NaN  NaN

In [63]:
record2.dropna(subset=['start'], how='any', axis=0, inplace=True)
len(record2)

395

In [57]:
## 예외 처리
'''
288 광고론, 290 매스컴이론, 323 웹드라마비평 에서
time을 물결~ 이 아닌 하이픈-으로 사용 중이라
컬럼 분리가 제대로 되지 않았음.
'''

indices_to_extract = [288, 290, 323]

# .loc[]를 사용하여 특정 인덱스에 해당하는 행만 선택합니다.
except_time = record2.loc[indices_to_extract]

# 하이픈으로 컬럼 분리
except_time['end'] = except_time.start.str.split('-').str[1]
except_time['start'] = except_time.start.str.split('-').str[0]

## --> 인덱스는 그대로 유지됨



In [58]:
except_time

kind     code classname prof room1 room2 date1 date2  start    end
288   전선  DJ00060       광고론  김인혜  7309   NaN     목   NaN  12:00  14:50
290   전선  DJ00062     매스컴이론  이준형  7308   NaN     목   NaN  09:00  11:50
323   전선  DP00008    웹드라마비평   강인  3501   NaN     목   NaN  09:00  11:50

In [59]:
# 처리한 프레임을 인덱스를 기준으로 원본 프레임에 적용
record2.update(except_time)

In [62]:
record2[287:289]

kind     code classname prof room1 room2 date1 date2  start    end
290   전선  DJ00062     매스컴이론  이준형  7308   NaN     목   NaN  09:00  11:50
291   전선  DP00006   영상제작실습2  연희승  M401   NaN     수   NaN  09:00  11:50

In [64]:
# room  없는 강의 삭제
record2[record2.room1.isna()]

Empty DataFrame
Columns: [kind, code, classname, prof, room1, room2, date1, date2, start, end]
Index: []

In [65]:
record2[record2['room1'] == '9999']

kind     code classname prof room1 room2 date1 date2  start    end
214   전선  SW00023  사회복지현장실습  서진환  9999   NaN     토   NaN  09:00  11:50
215   전선  SW00023  사회복지현장실습  이동림  9999   NaN     토   NaN  09:00  11:50

In [66]:
record2 = record2[record2['room1'] != '9999'].reset_index(drop=True)

In [67]:
record2[record2['room1'] == '9999']

Empty DataFrame
Columns: [kind, code, classname, prof, room1, room2, date1, date2, start, end]
Index: []

In [68]:
 record2[record2['room1'] =='0000']

Empty DataFrame
Columns: [kind, code, classname, prof, room1, room2, date1, date2, start, end]
Index: []

In [69]:
 record2[record2['room1'] =='미지정']

Empty DataFrame
Columns: [kind, code, classname, prof, room1, room2, date1, date2, start, end]
Index: []

In [70]:
 record2[record2['room1'] =='']

Empty DataFrame
Columns: [kind, code, classname, prof, room1, room2, date1, date2, start, end]
Index: []

In [71]:
 record2[record2['date1'] =='']

Empty DataFrame
Columns: [kind, code, classname, prof, room1, room2, date1, date2, start, end]
Index: []

In [73]:
len(record2)

393

## time이 콜론으로 분리되지 않은 것 찾기

In [74]:
len(record2[record2['start'].str.contains(":") & \
            record2['end'].str.contains(":")])

393

## 중복 강의 제거

In [88]:
record3 = record2.copy()

In [89]:
# 중복된 행만 모아보기
df_duplicates = record3[record3.duplicated(keep=False)]
df_duplicates

kind     code         classname      prof room1 room2 date1 date2  start  \
168   전선  HR00005  인공지능시대의종교와미디어의미래  최영묵\n이주엽  7306   NaN     목   NaN  15:00   
169   전선  HR00007           종교와사회복지  노철래\n한지수  7305   NaN     화   NaN  12:00   
220   전선  HR00007           종교와사회복지  노철래\n한지수  7305   NaN     화   NaN  12:00   
264   전선  DJ00001         스토리텔링과미디어       백정숙  3401   NaN     화   NaN  15:00   
270   전선  DJ00011          광고크리에이티브        윤목  M403   NaN     월   NaN  15:00   
281   전선  DJ00050       스포츠콘텐츠이해와실습       정윤수  7309   NaN     화   NaN  12:00   
321   전선  DJ00001         스토리텔링과미디어       백정숙  3401   NaN     화   NaN  15:00   
322   전선  DJ00011          광고크리에이티브        윤목  M403   NaN     월   NaN  15:00   
324   전선  DJ00050       스포츠콘텐츠이해와실습       정윤수  7309   NaN     화   NaN  12:00   
325   전선  HR00005  인공지능시대의종교와미디어의미래  최영묵\n이주엽  7306   NaN     목   NaN  15:00   
329   전필  FI00001            머신러닝입문       홍성준  6206   NaN     화   NaN  09:00   
330   전필  FI00003         데이터사이언스입문       이상윤  6206   NaN     목   NaN  09:00   
337   전선  XB00032              졸업지도       김명철  6206   NaN     목   NaN  18:25   
339   전필  XB00001              운영체제       노은하  6201  6201     화     목  16:30   
340   전필  XB00001              운영체제       노은하  6201  6201     화     목  16:30   
353   전선  IS00015          백엔드프로그래밍       김선형  6202  6202     화     목  15:00   
354   전선  IS00018         서버구축및형상관리       이승진  6202  6202     월     수  16:30   
355   전선  XC00026             인터넷보안       이종현  6406   NaN     금   NaN  12:00   
356   전선  XC00031            정보보호개론       이종현  6406   NaN     금   NaN  15:00   
360   전선  XC00026             인터넷보안       이종현  6406   NaN     금   NaN  12:00   
362   전선  XC00031            정보보호개론       이종현  6406   NaN     금   NaN  15:00   
363   전선  IS00014          JSP프로그래밍       박정식  6404   NaN     화   NaN  09:00   
364   전선  IS00006             데이터통신       박정식  6404  6404     화     목  15:00   
365   전선  XB00032              졸업지도       김명철  6206   NaN     목   NaN  18:25   
371   전필  FI00001            머신러닝입문       홍성준  6206   NaN     화   NaN  09:00   
372   전필  FI00003         데이터사이언스입문       이상윤  6206   NaN     목   NaN  09:00   
375   전선  IS00006             데이터통신       박정식  6404  6404     화     목  15:00   
384   전선  IS00014          JSP프로그래밍       박정식  6404   NaN     화   NaN  09:00   
385   전선  IS00015          백엔드프로그래밍       김선형  6202  6202     화     목  15:00   
390   전선  IS00018         서버구축및형상관리       이승진  6202  6202     월     수  16:30   

       end  
168  17:50  
169  14:50  
220  14:50  
264  17:50  
270  17:50  
281  14:50  
321  17:50  
322  17:50  
324  14:50  
325  17:50  
329  11:50  
330  11:50  
337  19:15  
339  17:45  
340  17:45  
353  16:15  
354  17:45  
355  14:50  
356  17:50  
360  14:50  
362  17:50  
363  11:50  
364  16:15  
365  19:15  
371  11:50  
372  11:50  
375  16:15  
384  11:50  
385  16:15  
390  17:45

### 중복 15개
### 전선의 경우 전공 공동 개설 등의 이유로 중복이 생기는 듯 하다 .
### 393 - 15 = 378

In [90]:
# keep='first'를 통해 중복된 행 중 첫번째 행에는 false가 저장됨.
# drop을 통해 중복이 있다면 false만 남기고 다 제거됨
df_deduplicated_direct = record3.drop_duplicates(keep='first')

In [91]:
df_deduplicated_direct.reset_index(drop=True, inplace=True)

In [94]:
len(df_deduplicated_direct)

378

In [93]:
df_deduplicated_direct[370:]

kind     code  classname prof room1 room2 date1 date2  start    end
370   전선  IS00013      컴퓨터구조  김학수  3301   NaN     수   NaN  09:00  11:50
371   전선  IS00013      컴퓨터구조  정인철  3301  3301     월     수  13:30  14:45
372   전선  IS00015   백엔드프로그래밍  김선형  6202  6202     화     목  16:30  17:45
373   전선  IS00016    컴퓨터네트워크  박정식  6406   NaN     목   NaN  09:00  11:50
374   전선  IS00017    디지털회로실험  정인철  6406   NaN     월   NaN  09:00  11:50
375   전선  IS00018  서버구축및형상관리  이승진  6202  6202     월     수  15:00  16:15
376   전선  IS00019     신호및시스템  김학수  6201  6201     월     수  13:30  14:45
377   전선  IS00020       웹디자인  이해신  3302   NaN     금   NaN  09:00  11:50

# 최종 데이터 프레임

In [95]:
final_df = df_deduplicated_direct.copy()

In [96]:
final_df.head()

kind     code             classname prof room1 room2 date1 date2  start  \
0   교필  AC00007   인권과평화(주제:식민주의와인종주의)  조경희  7207   NaN   화요일   NaN  10:00   
1   교필  AC00007       인권과평화(주제:전쟁과평화)  강성현  7308   NaN   화요일   NaN  10:00   
2   교필  AC00007  인권과평화(주제:세계시민의비전과도전)  이상윤  7208   NaN   화요일   NaN  10:00   
3   교필  AC00007          인권과평화(주제:미정)  오영숙  3502   NaN   화요일   NaN  10:00   
4   교필  AC00013                   말과글  오현화  1503   NaN   수요일   NaN  09:00   

     end  
0  11:50  
1  11:50  
2  11:50  
3  11:50  
4  10:50

In [97]:
final_df.tail()

kind     code  classname prof room1 room2 date1 date2  start    end
373   전선  IS00016    컴퓨터네트워크  박정식  6406   NaN     목   NaN  09:00  11:50
374   전선  IS00017    디지털회로실험  정인철  6406   NaN     월   NaN  09:00  11:50
375   전선  IS00018  서버구축및형상관리  이승진  6202  6202     월     수  15:00  16:15
376   전선  IS00019     신호및시스템  김학수  6201  6201     월     수  13:30  14:45
377   전선  IS00020       웹디자인  이해신  3302   NaN     금   NaN  09:00  11:50

In [98]:
len(final_df)

378

In [99]:
final_df.to_csv('csv25_2.csv', index=False, encoding='utf-8-sig')